File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/PaymentDetails.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class PaymentDetails : ISerialized
    {
        public PaymentDetails(XmlNode xmlNode)
        {
            XmlNode paymentReferenceNode = xmlNode.SelectSingleNode("paymentReference");
            
            if (paymentReferenceNode != null)
            {
                if (paymentReferenceNode.Attributes["href"] != null || paymentReferenceNode.Attributes["id"] != null) 
                {
                    if (paymentReferenceNode.Attributes["id"] != null) 
                    {
                        paymentReferenceIDRef_ = paymentReferenceNode.Attributes["id"].Value;
                        PaymentReference ob = new PaymentReference(paymentReferenceNode);
                        IDManager.SetID(paymentReferenceIDRef_, ob);
                    }
                    else if (paymentReferenceNode.Attributes["href"] != null)
                    {
                        paymentReferenceIDRef_ = paymentReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        paymentReference_ = new PaymentReference(paymentReferenceNode);
                    }
                }
                else
                {
                    paymentReference_ = new PaymentReference(paymentReferenceNode);
                }
            }
            
        
            XmlNodeList grossCashflowNodeList = xmlNode.SelectNodes("grossCashflow");
            
            if (grossCashflowNodeList != null)
            {
                this.grossCashflow_ = new List<GrossCashflow>();
                foreach (XmlNode item in grossCashflowNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            grossCashflowIDRef_ = item.Attributes["id"].Value;
                            grossCashflow_.Add(new GrossCashflow(item));
                            IDManager.SetID(grossCashflowIDRef_, grossCashflow_[grossCashflow_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            grossCashflowIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        grossCashflow_.Add(new GrossCashflow(item));
                        }
                    }
                    else
                    {
                        grossCashflow_.Add(new GrossCashflow(item));
                    }
                }
            }
            
        
            XmlNode settlementInformationNode = xmlNode.SelectSingleNode("settlementInformation");
            
            if (settlementInformationNode != null)
            {
                if (settlementInformationNode.Attributes["href"] != null || settlementInformationNode.Attributes["id"] != null) 
                {
                    if (settlementInformationNode.Attributes["id"] != null) 
                    {
                        settlementInformationIDRef_ = settlementInformationNode.Attributes["id"].Value;
                        SettlementInformation ob = new SettlementInformation(settlementInformationNode);
                        IDManager.SetID(settlementInformationIDRef_, ob);
                    }
                    else if (settlementInformationNode.Attributes["href"] != null)
                    {
                        settlementInformationIDRef_ = settlementInformationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementInformation_ = new SettlementInformation(settlementInformationNode);
                    }
                }
                else
                {
                    settlementInformation_ = new SettlementInformation(settlementInformationNode);
                }
            }
            
        
        }
        
    
        #region PaymentReference_
        private PaymentReference paymentReference_;
        public PaymentReference PaymentReference_
        {
            get
            {
                if (this.paymentReference_ != null)
                {
                    return this.paymentReference_; 
                }
                else if (this.paymentReferenceIDRef_ != null)
                {
                    paymentReference_ = IDManager.getID(paymentReferenceIDRef_) as PaymentReference;
                    return this.paymentReference_; 
                }
                else
                {
                      return this.paymentReference_; 
                }
            }
            set
            {
                if (this.paymentReference_ != value)
                {
                    this.paymentReference_ = value;
                }
            }
        }
        #endregion
        
        public string paymentReferenceIDRef_ { get; set; }
        #region GrossCashflow_
        private List<GrossCashflow> grossCashflow_;
        public List<GrossCashflow> GrossCashflow_
        {
            get
            {
                if (this.grossCashflow_ != null)
                {
                    return this.grossCashflow_; 
                }
                else if (this.grossCashflowIDRef_ != null)
                {
                    return this.grossCashflow_; 
                }
                else
                {
                      return this.grossCashflow_; 
                }
            }
            set
            {
                if (this.grossCashflow_ != value)
                {
                    this.grossCashflow_ = value;
                }
            }
        }
        #endregion
        
        public string grossCashflowIDRef_ { get; set; }
        #region SettlementInformation_
        private SettlementInformation settlementInformation_;
        public SettlementInformation SettlementInformation_
        {
            get
            {
                if (this.settlementInformation_ != null)
                {
                    return this.settlementInformation_; 
                }
                else if (this.settlementInformationIDRef_ != null)
                {
                    settlementInformation_ = IDManager.getID(settlementInformationIDRef_) as SettlementInformation;
                    return this.settlementInformation_; 
                }
                else
                {
                      return this.settlementInformation_; 
                }
            }
            set
            {
                if (this.settlementInformation_ != value)
                {
                    this.settlementInformation_ = value;
                }
            }
        }
        #endregion
        
        public string settlementInformationIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/PaymentDetails.cpp
// PaymentDetails.cpp 
#include "PaymentDetails.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

PaymentDetails::PaymentDetails(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //paymentReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* paymentReferenceNode = xmlNode->FirstChildElement("paymentReference");

   if(paymentReferenceNode){paymentReferenceIsNull_ = false;}
   else{paymentReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- paymentReferenceNode , address : " << paymentReferenceNode << std::endl;
   #endif
   if(paymentReferenceNode)
   {
      if(paymentReferenceNode->Attribute("href") || paymentReferenceNode->Attribute("id"))
      {
          if(paymentReferenceNode->Attribute("id"))
          {
             paymentReferenceIDRef_ = paymentReferenceNode->Attribute("id");
             paymentReference_ = boost::shared_ptr<PaymentReference>(new PaymentReference(paymentReferenceNode));
             paymentReference_->setID(paymentReferenceIDRef_);
             IDManager::instance().setID(paymentReferenceIDRef_,paymentReference_);
          }
          else if(paymentReferenceNode->Attribute("href")) { paymentReferenceIDRef_ = paymentReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { paymentReference_ = boost::shared_ptr<PaymentReference>(new PaymentReference(paymentReferenceNode));}
   }

   //grossCashflowNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* grossCashflowNode = xmlNode->FirstChildElement("grossCashflow");

   if(grossCashflowNode){grossCashflowIsNull_ = false;}
   else{grossCashflowIsNull_ = true;}

   if(grossCashflowNode)
   {
      for(grossCashflowNode; grossCashflowNode; grossCashflowNode = grossCashflowNode->NextSiblingElement("grossCashflow")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- grossCashflowNode , address : " << grossCashflowNode << std::endl;
          #endif
          if(grossCashflowNode->Attribute("href") || grossCashflowNode->Attribute("id"))
          {
              if(grossCashflowNode->Attribute("id"))
              {
                  grossCashflowIDRef_ = grossCashflowNode->Attribute("id");
                  grossCashflow_.push_back(boost::shared_ptr<GrossCashflow>(new GrossCashflow(grossCashflowNode)));
                  grossCashflow_.back()->setID(grossCashflowIDRef_);
                  IDManager::instance().setID(grossCashflowIDRef_, grossCashflow_.back());
              }
              else if(grossCashflowNode->Attribute("href")) { grossCashflowIDRef_ = grossCashflowNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { grossCashflow_.push_back(boost::shared_ptr<GrossCashflow>(new GrossCashflow(grossCashflowNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- grossCashflowNode , address : " << grossCashflowNode << std::endl;
       #endif
   }

   //settlementInformationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementInformationNode = xmlNode->FirstChildElement("settlementInformation");

   if(settlementInformationNode){settlementInformationIsNull_ = false;}
   else{settlementInformationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementInformationNode , address : " << settlementInformationNode << std::endl;
   #endif
   if(settlementInformationNode)
   {
      if(settlementInformationNode->Attribute("href") || settlementInformationNode->Attribute("id"))
      {
          if(settlementInformationNode->Attribute("id"))
          {
             settlementInformationIDRef_ = settlementInformationNode->Attribute("id");
             settlementInformation_ = boost::shared_ptr<SettlementInformation>(new SettlementInformation(settlementInformationNode));
             settlementInformation_->setID(settlementInformationIDRef_);
             IDManager::instance().setID(settlementInformationIDRef_,settlementInformation_);
          }
          else if(settlementInformationNode->Attribute("href")) { settlementInformationIDRef_ = settlementInformationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementInformation_ = boost::shared_ptr<SettlementInformation>(new SettlementInformation(settlementInformationNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<PaymentReference> PaymentDetails::getPaymentReference()
{
   if(!this->paymentReferenceIsNull_){
        if(paymentReferenceIDRef_ != ""){
             return boost::shared_static_cast<PaymentReference>(IDManager::instance().getID(paymentReferenceIDRef_));
        }else{
             return this->paymentReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PaymentReference>();
   }
}
std::vector<boost::shared_ptr<GrossCashflow>> PaymentDetails::getGrossCashflow()
{
   if(!this->grossCashflowIsNull_){
        if(grossCashflowIDRef_ != ""){
             return this->grossCashflow_;
        }else{
             return this->grossCashflow_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<GrossCashflow>>();
   }
}
boost::shared_ptr<SettlementInformation> PaymentDetails::getSettlementInformation()
{
   if(!this->settlementInformationIsNull_){
        if(settlementInformationIDRef_ != ""){
             return boost::shared_static_cast<SettlementInformation>(IDManager::instance().getID(settlementInformationIDRef_));
        }else{
             return this->settlementInformation_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SettlementInformation>();
   }
}
}

